src/kmp.h

```diff
@@ -0,0 +1,68 @@
+#ifndef KMP_H
+#define KMP_H
+
+#include <malloc.h>
+
+int *compute_border_array(char *pattern, int pattern_len) {
+    if (pattern_len == 0) return NULL;
+    int *ba = malloc(sizeof(int)*pattern_len);
+    ba[0] = 0;
+    for (int i = 1; i < pattern_len; i++) {
+        // Test if we can extend border. Otherwise, reset.
+        if (pattern[i] == pattern[ba[i-1]]) {
+            ba[i] = ba[i-1]+1;
+        } else {
+            ba[i] = 0;
+        }
+    }
```

Comment:
This doesn't compute the correct border array. If you want the longest border, it is not either an extension of the previous string or zero. It can be an extension of a border of a border. This will not guarantee a linear time algorithm (although it will work fine for random strings or strings that are all the same letter since there you can always extend).